[ip] Correlations as normalized conditional entropy

minafarid opened the pull request:
[Do not merge yet]

* Implemented normalized conditional entropy as a measure of correlation.
* Modified some learning parameters in example

Current performance:
```
Precision = 1.00, Recall = 0.67, Repairing Recall = 0.79, 
F1 = 0.80, Repairing F1 = 0.88, 
Detected Errors = 435, Total Errors = 509, 
Correct Repairs = 342, Total Repairs = 342, Total Repairs (Grdth present) = 342
```

josmcg commented:
@j48zheng I haven't used spark a ton, do you know what function I would use to do that?

minafarid committed:
Correlations as normalized conditional entropy

minafarid committed:
Minor optimization in correlations

minafarid committed:
normalize conditional entropy by log base

Current performance:
```
Precision = 0.96, Recall = 0.85, Repairing Recall = 0.99,
F1 = 0.90, Repairing F1 = 0.97,
Detected Errors = 435, Total Errors = 509,
Correct Repairs = 431, Total Repairs = 450, Total Repairs (Grdth present) = 450
```

richardwu commented on examples/holoclean_repair_example.py:
```diff
@@ -37,8 +38,8 @@
 featurizers = [
-    InitAttrFeaturizer(),
-    InitSimFeaturizer(),
+    InitAttrFeaturizer(learnable=False),
```
Are we fixing these in this commit? Maybe you should do it in a separate commit.

minafarid replied:
Agreed.. Reverted

richardwu commented on domain/domain.py:
```diff
@@ -40,29 +42,68 @@ def setup(self):
+                # If H(x|y) = 0, then y determines x, i.e., y -> x.
+                # Use the domain size of x as a log base for normalization.
+                y_vals = data_df[y]
+                x_y_entropy = drv.entropy_conditional(x_vals, y_vals, base=x_domain_size)
```
Hmmm, I'm not too sure about introducing a new package dependency here just for conditional entropy, considering the project doesn't seem to be very active.

Implementing conditional entropy here with MLE shouldn't be too difficult considering that `x_vals` and `y_vals` are stored into a DataFrame anyways.

minafarid replied:
This is a first iteration to prove that conditional entropy works well for our needs. The project was last updated 2 months ago, so it is not very old. 
I agree that a better thing to do is to implement our own conditional entropy, but this was faster to implement. We can revisit this in the next dev meeting.

minafarid committed:
Reverting example
